src/kokkos/KokkosCore/HistoContainer.h

```diff
@@ -0,0 +1,339 @@
+    for (uint32_t i = 0; i < totbins(); ++i) {
+      Kokkos::atomic_fetch_add(off + i, co(0).off[i]);
+    }
+  }
+
+  // template <typename HistoType>
+  // KOKKOS_INLINE_FUNCTION static void add(Kokkos::View<HistoType*,KokkosExecSpace> const histo,Kokkos::View<CountersOnly*,KokkosExecSpace> const co){
+  //   Kokkos::parallel_for("histo_add",Kokkos::RangePolicy<KokkosExecSpace>(0,totbins()),
+  //     KOKKOS_LAMBDA(const int& i){
+  //       Kokkos::atomic_fetch_add(histo(0).off + i,co(0).off[i]);
+  //     });
+  // }
+
+  static KOKKOS_INLINE_FUNCTION uint32_t atomicIncrement(Counter &x) {
+    return Kokkos::atomic_fetch_add(&x, 1);
+  }
+
+  static KOKKOS_INLINE_FUNCTION uint32_t atomicDecrement(Counter &x) {
+    return Kokkos::atomic_fetch_sub(&x, 1);
+  }
+
+  KOKKOS_INLINE_FUNCTION void countDirect(T b) {
+    assert(b < nbins());
+    atomicIncrement(off[b]);
+  }
+
+  KOKKOS_INLINE_FUNCTION void fillDirect(T b, index_type j) {
+    assert(b < nbins());
+    auto w = atomicDecrement(off[b]);
+    assert(w > 0);
+    bins[w - 1] = j;
+  }
+
+  KOKKOS_INLINE_FUNCTION int32_t bulkFill(Kokkos::View<AtomicPairCounter*,KokkosExecSpace> apc, index_type const *v, uint32_t n) {
+    auto c = apc(0).add(n);
+    if (c.m >= nbins())
+      return -int32_t(c.m);
+    off[c.m] = c.n;
+    for (uint32_t j = 0; j < n; ++j)
+      bins[c.n + j] = v[j];
+    return c.m;
+  }
+
+  KOKKOS_INLINE_FUNCTION void bulkFinalize(Kokkos::View<AtomicPairCounter*,KokkosExecSpace> const apc) {
+    off[apc(0).get().m] = apc(0).get().n;
+  }
+
+  KOKKOS_INLINE_FUNCTION void bulkFinalizeFill(Kokkos::View<AtomicPairCounter*,KokkosExecSpace> const apc,const int& threadId) {
+    auto m = apc(0).get().m;
+    auto n = apc(0).get().n;
+    if (m >= nbins()) {  // overflow!
+      off[nbins()] = uint32_t(off[nbins() - 1]);
+      return;
+    }
+    auto i = m + threadId;
+    if(i < totbins()){
+      off[i] = n;
+    }
+  }
+
+  KOKKOS_INLINE_FUNCTION void count(T t) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    atomicIncrement(off[b]);
+  }
+
+  KOKKOS_INLINE_FUNCTION void fill(T t, index_type j) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    auto w = atomicDecrement(off[b]);
+    assert(w > 0);
+    bins[w - 1] = j;
+  }
+
+  KOKKOS_INLINE_FUNCTION void count(T t, uint32_t nh) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    b += histOff(nh);
+    assert(b < totbins());
+    atomicIncrement(off[b]);
+  }
+
+  KOKKOS_INLINE_FUNCTION void fill(T t, index_type j, uint32_t nh) {
+    uint32_t b = bin(t);
+    assert(b < nbins());
+    b += histOff(nh);
+    assert(b < totbins());
+    auto w = atomicDecrement(off[b]);
+    assert(w > 0);
+    bins[w - 1] = j;
+  }
+
+  #pragma hd_warning_disable
+  template<typename Histo>
+  static KOKKOS_INLINE_FUNCTION void finalize(Kokkos::View<Histo*,KokkosExecSpace> histo) {
```

Comment:
Is this function supposed to be called only from host? In that case I belive the `KOKKOS_INLINE_FUNCTION` would better be dropped (I'd also consider making this a free function, but am not feeling strongly about it).

Do I understand/recall correctly that the difference wrt. CUDA code is that the (handcrafted)  `blockPrefixScan()` is designed to be called from device code, whereas Kokkos' `parallel_scan()` (in this case) needs to be invoked from host?